File: 프로그래머스/lv2/87946. 피로도/피로도.java
import java.util.ArrayList;
import java.util.Arrays;

class Solution {
	static int answer = 0;
	static int N;
    public int solution(int k, int[][] dungeons) {
    	N = dungeons.length;
    	ArrayList<int[]> list = new ArrayList<>();
    	for (int i = 0; i < N; i++) {
			list.add(dungeons[i]);
		}
    	
    	
    	boolean[] visited = new boolean[N];
    	dfs(0, k, visited, dungeons);
    	
        return answer;
    }
    public static void dfs(int depth, int k, boolean[] visited, int[][] dungeons) {
    	
    	for (int i = 0; i < N; i++) {
			if (visited[i]) continue;
			if (k >= dungeons[i][0]) {
				visited[i] = true;
				dfs(depth+1, k-dungeons[i][1], visited, dungeons);
				visited[i] = false;
			}
		}
    	
    	answer = Math.max(answer, depth);
    	return;
    	
    }
}

File: 프로그래머스/2/87946. 피로도/피로도.js
function solution(k, dungeons) {
  var answer = -1;

  const visited = Array(dungeons.length).fill(false);

  function dfs(currentHp, depth) {
    answer = Math.max(answer, depth);

    for (let i = 0; i < dungeons.length; i++) {
      const [minHp, useHp] = dungeons[i];
      if (visited[i]) continue;
      if (currentHp >= minHp) {
        visited[i] = true;
        dfs(currentHp - useHp, depth + 1);
        visited[i] = false;
      }
    }
  }

  dfs(k, 0);

  return answer;
}
